includes/user-functions.php

```diff
@@ -444,6 +444,47 @@ function give_validate_user_password( $password = '', $confirm_password = '', $r
 
 
 /**
+ * Looks up donations by email that match the registering user.
+ *
+ * This is for users that donated as a guest and then came back and created an account.
+ *
+ * @param int $user_id The new user's ID.
+ *
+ * @access public
+ * @since 1.0
+ *
+ * @return void
+ */
+function give_add_past_donations_to_new_user( $user_id ) {
+
+	$email = get_the_author_meta( 'user_email', $user_id );
+
+	$payments = give_get_payments( array(
+		's' => $email,
+	) );
+
+	if ( $payments ) {
+		foreach ( $payments as $payment ) {
+			if ( intval( give_get_payment_user_id( $payment->ID ) ) > 0 ) {
+				continue;
+			}
+
+			$meta                    = give_get_payment_meta( $payment->ID );
+			$meta['user_info']       = maybe_unserialize( $meta['user_info'] );
+			$meta['user_info']['id'] = $user_id;
+
+			// Store the updated user ID in the payment meta.
+			give_update_payment_meta( $payment->ID, '_give_payment_meta', $meta );
+			give_update_payment_meta( $payment->ID, '_give_payment_user_id', $user_id );
+		}
+	}
+
+}
+
+add_action( 'user_register', 'give_add_past_donations_to_new_user' );
```

Comment:
@mehul0810 Remove this code because payments now connect with donor, so if user change or update it will not going to affect anything